src/coreclr/src/jit/morph.cpp

```diff
@@ -11611,6 +11611,68 @@ GenTree* Compiler::fgMorphFieldAssignToSIMDIntrinsicSet(GenTree* tree)
 
 #endif // FEATURE_SIMD
 
+//------------------------------------------------------------------------------
+// fgMorphCommutative : Try to simplify "(X op C1) op C2" to "X op (C1 op C2)"
+//                      for commutative operators.
+//
+// Arguments:
+//       tree - node to fold
+//
+// return value:
+//       A folded GenTree* instance or nullptr if something prevents folding.
+//
+
+GenTree* Compiler::fgMorphCommutative(GenTreeOp* tree)
+{
+    assert(varTypeIsIntegralOrI(tree->TypeGet()));
+    assert(tree->OperIs(GT_ADD, GT_MUL, GT_OR, GT_AND, GT_XOR));
+
+    // op1 can be GT_COMMA, in this case we're going to fold
+    // "COMMA(..., (X op C1)) op C2" to "COMMA(..., X op (C1 op C2))"
```

Comment:
Is this comment stale? It doesn't look like you handle the COMMA case.

Not sure it's worth the trouble. Might be a something to try and measure.

Reply:
hm it does handle COMMA for me, it has `gtEffectiveVal(true)` means if it's COMMA - take gtOp2 of it
and preserves it in the end, will post here an example

Reply:
Maybe it's just the comment is misleading, it should be (in prefix notation)
```
(OP (COMMA(... (OP x C1))) C2)
==>
(COMMA(... (OP x (OP C1 C2))))
```
```

Reply:
will change the comment, here is the case:

![image](https://user-images.githubusercontent.com/523221/96774685-04964300-13ef-11eb-83b3-e8a86e692812.png)

Reply:
Good. Sorry for the confusion.